modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -3113,8 +3117,7 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
 
 	GDScriptParser::ClassNode *base_class = base.class_type;
 
-	// TODO: Switch current class/function/suite here to avoid misrepresenting identifiers (in recursive reduce calls).
-	while (base_class != nullptr) {
+	if (base_class != nullptr) {
 		if (base_class->identifier && base_class->identifier->name == name) {
 			p_identifier->set_datatype(base_class->get_datatype());
 			return;
```

Comment:
I tested this and it did not seem to fix the issue you mentionned.

Reply:
To fix it you need to add `p_base == nullptr` check. It does not make sense to type access to the class names through class itself (`A.A` or `B.A`) or class instance (`a.A` or `b.A`) (this will lead to runtime error anyway). Only when there is no base (so simply `A`) it can be typed properly. And it should have your constant logic so that you can do `const X := A` inside class itself, for alias for example.